NanoMips: OR combine optimization

nikolaperic opened the pull request:
Combine OR into INS when there is no masking for
the second register but it holds zero extended
value whose original width is less than equal to
the number of bits from the first register set
to zero.

This resolves the following issue:

> Bitfield insert a value that can fit into value such as 0xff0004 (do not assume it is exactly the same width).

Commit:
NanoMips: OR combine optimization

Combine OR into INS when there is no masking for
the second register but it holds zero extended
value whose original width is less than equal to
the number of bits from the first register set
to zero.

Commit:
NanoMips: Support for 16bit masks in OR to INS combine

This allows the INS instruction to be used to set the
value of one field of the bitfields struct.

nikolaperic commented:
Added one more commit which allows 16-bit masks to be recognized as valid inside OR to INS combiner. This makes it possible to use INS instruction to set the value of one field of the bitfields struct. Setting multiple fields consecutively using the INS instruction is not yet supported.

cme commented on llvm/lib/Target/Mips/MipsISelLowering.cpp:
```diff
@@ -1046,8 +1046,26 @@ static SDValue attemptOrToIns(SDValue &And, SDValue &Right, SDNode *N,
-  if (!isShiftedMask(Mask0, SMPos0, SMSize0))
-    return SDValue();
+
+  KnownBits KnownMaskBits = DAG.computeKnownBits(And.getOperand(1));
```
Can use `CN` instead of `And.getOperand(1)` here, which should be easier to read too.

nikolaperic replied:
Method ```DAG.computeKnownBits()``` requires an argument of type ```SDValue```, that's the reason why ```And.getOperand(1)``` is passed instead of ```CN```. Maybe we can somehow extract ```SDValue``` from ```CN```?

cme replied:
Oh, I didn't spot that because the conversion the other way from SDValue to SDNode is implicit. Yes, you should be able to get the SDValue corresponding to a SDNode's *n*'th result just by calling SDValue(node, *n*).

nikolaperic replied:
Thanks! I'll amend this in last commit.

cme commented on llvm/lib/Target/Mips/MipsISelLowering.cpp:
```diff
@@ -1046,8 +1046,26 @@ static SDValue attemptOrToIns(SDValue &And, SDValue &Right, SDNode *N,
+
+  KnownBits KnownMaskBits = DAG.computeKnownBits(And.getOperand(1));
+  APInt maxValue = KnownMaskBits.getMaxValue();
+  unsigned int MaxNumOfActiveBitsInMask = maxValue.getActiveBits();
```
Could replace this with `KnownBits::countActiveBits()`?

nikolaperic replied:
Did you mean ```KnownBits::countMaxActiveBits()```? 
But still, ```KnownBits::countMaxActiveBits()``` does not exist in our version of LLVM. Maybe we can ```cherry-pick``` the commit that introduced it?

cme replied:
Ah! Yes, I did mean `countMaxActiveBits` and indeed didn't spot that it's not in '13. Not worth a cherry-pick just to clean that up so I'll withdraw this :)

cme commented on llvm/lib/Target/Mips/MipsISelLowering.cpp:
```diff
@@ -1129,9 +1147,20 @@ static SDValue attemptOrToIns(SDValue &And, SDValue &Right, SDNode *N,
+      // then we will allow replacement with the INS instruction
+      // because we are sure that we have only leading zeros
+      KnownBits Known = DAG.computeKnownBits(Right.getOperand(0));
+      if (Known.countMaxPopulation() > SMSize0)
```
This seems wrong. countMaxPopulation() returns the count of the number of bits that could be one, the maximum possible value of `popcnt()`, without saying anything about *where* those 1 bits are. Don't we need to make sure here that they're only in the lowest SMSize0 bits?

nikolaperic replied:
Yes, I agree, we don't have that guarantee with the current solution. Maybe it is enough to check if ```Known.getMaxValue().getActiveBits()``` is smaller than ```SMSize0``` and to allow ```INS``` to be generated in that case? I think that should give us maximum number of bits set to 1 starting from the lowest positions.

nikolaperic replied:
I have uploaded the proposed change.

cme replied:
Yes, I think `getActiveBits` is the right thing here.

djtodoro commented:
@nikolaperic should be created against `nanomips` branch.

Commit:
NanoMips: OR combine fix

Allow OR to INS combine without second register
masking only if maximum possible value given by
KnownBits for the second register can fit
inside SMSize0.

draganmladjenovic commented:
@cme Is there anything left blocking us from merging this?

cme commented:
> @cme Is there anything left blocking us from merging this?

No, it has already been merged
